File: go/thompson/linear.go
package thompson

// LinearWeights is a stub for linear contextual bandit.
type LinearWeights struct {
	Weights []float64
}

func NewLinearWeights(dim int) *LinearWeights {
	return &LinearWeights{Weights: make([]float64, dim)}
}

func (lw *LinearWeights) Score(features []float64) float64 {
	sum := 0.0
	for i, w := range lw.Weights {
		if i < len(features) {
			sum += w * features[i]
		}
	}
	if sum < 0 {
		return 0
	}
	if sum > 1 {
		return 1
	}
	return sum
}

// LinearPolicy wraps Posterior with shared linear weights.
type LinearPolicy struct {
	Dim     int
	Weights *LinearWeights
}

func NewLinearPolicy(dim int) *LinearPolicy {
	return &LinearPolicy{Dim: dim, Weights: NewLinearWeights(dim)}
}

func (lp *LinearPolicy) AdjustedMean(p Posterior, features []float64) float64 {
	base := p.Mean()
	ctx := lp.Weights.Score(features)
	v := base*0.7 + ctx*0.3
	if v < 0 {
		return 0
	}
	if v > 1 {
		return 1
	}
	return v
}

File: crates/thompson-sampling/src/linear.rs
//! Linear contextual scaffold — future evolution of `PartitionedPolicy`.
//!
//! `PartitionedPolicy` (context.rs:33) buckets by `Context::partition_key()`.
//! Linear contextual extends this to share strength across contexts via
//! `Posterior` weighting. Stubbed here so thin waist stays `select`/`record`
//! without breaking `thompson-sampling` 0.1.x semver.

use crate::posterior::Posterior;

/// Weight vector for linear contextual bandit — placeholder.
#[derive(Debug, Clone)]
pub struct LinearWeights {
    pub weights: Vec<f64>,
}

impl LinearWeights {
    pub fn new(dim: usize) -> Self {
        Self {
            weights: vec![0.0; dim],
        }
    }

    /// Score for context features — dot product stub.
    pub fn score(&self, features: &[f64]) -> f64 {
        self.weights
            .iter()
            .zip(features)
            .map(|(w, f)| w * f)
            .sum::<f64>()
            .clamp(0.0, 1.0)
    }
}

/// Linear contextual policy stub — wraps `Posterior` per arm with shared weights.
#[derive(Debug)]
pub struct LinearPolicy {
    pub dim: usize,
    pub weights: LinearWeights,
}

impl LinearPolicy {
    pub fn new(dim: usize) -> Self {
        Self {
            dim,
            weights: LinearWeights::new(dim),
        }
    }

    /// Posterior mean adjusted by linear score — future hook for `select`.
    pub fn adjusted_mean(&self, posterior: &Posterior, features: &[f64]) -> f64 {
        let base = posterior.mean();
        let ctx = self.weights.score(features);
        (base * 0.7 + ctx * 0.3).clamp(0.0, 1.0)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::posterior::Posterior;

    #[test]
    fn adjusted_mean_within_bounds() {
        let policy = LinearPolicy::new(2);
        let post = Posterior::new(8.0, 2.0).unwrap();
        let m = policy.adjusted_mean(&post, &[0.5, 0.5]);
        assert!((0.0..=1.0).contains(&m));
    }
}
